tools/run_tests/sanity/cpp_banned_constructs.sh

```diff
@@ -0,0 +1,28 @@
+#!/bin/sh
+# Copyright 2019 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+
+cd "$(dirname "$0")/../../.."
+
+#
+# Prevent the use of synchronization constructs from std:: since the code should
+# be using grpc_core::Mutex, grpc::internal::Mutex, and related functions
+#
+
+egrep -Irn \
+    'std::(mutex( |$)|condition_variable( |$)|lock_guard<|unique_lock<)' \
```

Comment:
`<` and `( |$)` should be kind of doing the same thing you're trying to do, no?

Reply:
< is not a special character for regex processing - it's the end of the token since those are templated types. The other ones are not templated types, so the end of the token is whitespace (and clang-format makes sure that the only whitespace we have is space or newline) or the start of a comment (not allowed by clang-format without intervening whitespace).

Reply:
Actually, looking it over, I think the trying to detect tokenization thing was just a bad idea. Someone could always make a `std::unique_ptr<std::mutex>` and that would slip through. So I've just changed it to stop trying to do that since at least at the current time, there are no allowed members of `std::` that have these names as prefixes anyway. If such a thing arises in the future, we can change the script then.

Reply:
Also added thread since that shouldn't be used either (not just because we have our own but also because it's inherently race-prone)

Reply:
Oh, I see. < in extended regex GNU means "end of word" :D I thought you were doing this.